### PR TITLE
[c10d] Start deprecating *_multigpu APIs

### PR DESCRIPTION
**Pull request by kwen2501**
### Deprecation reasons:
- For most users training is on one GPU per process so these APIs are rarely used 
- They added one more API dimension
- They can be expressed in a composed manner
- They are not abstracted – specific to GPU
- They caused backend APIs and implementations to have nested `std::vector<std::vector<Tensor>>`, which is hard to read or maintain

### CODE REVIEW

**Comment by kwen2501** [comment]
The "CUDA out of memory" CI error is unrelated

**Review by XilunWu — APPROVED** [review]
lgtm. thx for working on it!

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -1328,6 +1328,11 @@ def broadcast_multigpu(tensor_list, src, group=None, async_op=False, src_tensor=
         None, if not async_op or if not part of the group
 
     """
+    warnings.warn(
+        "torch.distributed.broadcast_multigpu will be deprecated. If you must "
```
Comment (H-Huang) [inline torch/distributed/distributed_c10d.py:1332]: link to docs? (https://pytorch.org/docs/master/distributed.html#collective-functions)

**Review by H-Huang — APPROVED** [review]

File: torch/distributed/distributed_c10d.py (hunk 1 above)
Comment (kwen2501) [reply torch/distributed/distributed_c10d.py:1332]: Thanks. Added

**Pull request closed without merging**